enterprise/server/ociregistry/ociregistry.go

```diff
@@ -0,0 +1,305 @@
+package ociregistry
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+const (
+	rangeHeaderBytesPrefix = "bytes="
+
+	registryInstanceName = "registry_artifacts"
+)
+
+var (
+	manifestReqRE = regexp.MustCompile("/v2/(.+?)/manifests/(.+)")
+	blobReqRE     = regexp.MustCompile("/v2/(.+?)/blobs/(.+)")
+	blobUploadRE  = regexp.MustCompile("/v2/(.+?)/blobs/uploads/")
+
+	enableRegistry = flag.Bool("ociregistry.enabled", false, "Whether to enable registry services")
+)
+
+type registry struct {
+	env environment.Env
+}
+
+func Register(env environment.Env) error {
+	if !*enableRegistry {
+		return nil
+	}
+
+	mux := env.GetInternalHTTPMux()
```

Comment:
we should figure out if this is the right place for this -- @vadimberezniker and @bduffany probably have ideas

Reply:
I thought we talked about putting this in the Cache Proxy, no?

If so, that'd mean hooking it up somewhere in here: https://github.com/buildbuddy-io/buildbuddy/blob/4987ddd8f86898a2380925c4be08a8727a2f4f33/enterprise/server/cmd/cache_proxy/cache_proxy.go#L51

Reply:
in addition to moving it to the cache proxy, I think it'd be clearer to move the `mux.Handle()` call to `main()` rather than hiding it in `Register()` - that way all of the supported HTTP routes are visible in one place. (Our server setup in `libmain` has all the routes in one place, too)

Reply:
Agree we will eventually want it (only?) in the cache proxy, but because we don't run a proxy in us-west1, it seems  good to start with it in the app. It can even be in both places honestly, it's the same code. I don't feel like it's any work to move since it's just call to Register()?

Iain is separately dealing with some cache proxy performance issues and frequently turning it on and off, and I didn't want to tie the deployment of the oci registry mirror to that.

Reply:
I've gone ahead and registered OCI registry (that's a mouthful) in `libmain`. When you can, please take a look & verify it's all kosher.